web3/providers/base.py

```diff
@@ -47,44 +72,42 @@ class BaseProvider:
     global_ccip_read_enabled: bool = True
     ccip_read_max_redirects: int = 4
 
-    @property
-    def middlewares(self) -> Tuple[Middleware, ...]:
-        return self._middlewares
+    # request caching
+    cache_allowed_requests: bool = False
+    cacheable_requests: Set[RPCEndpoint] = CACHEABLE_REQUESTS
+    _request_cache: SimpleCache
+    _request_cache_lock: threading.Lock = threading.Lock()
 
-    @middlewares.setter
-    def middlewares(self, values: MiddlewareOnion) -> None:
-        # tuple(values) converts to MiddlewareOnion -> Tuple[Middleware, ...]
-        self._middlewares = tuple(values)  # type: ignore
+    def __init__(self) -> None:
+        self._request_cache = SimpleCache(1000)
 
     def request_func(
-        self, w3: "Web3", outer_middlewares: MiddlewareOnion
+        self, w3: "Web3", middlewares: MiddlewareOnion
     ) -> Callable[..., RPCResponse]:
         """
-        @param outer_middlewares is an iterable of middlewares,
+        @param w3 is the web3 instance
+        @param middlewares is an iterable of middlewares,
             ordered by first to execute
         @returns a function that calls all the middleware and
             eventually self.make_request()
         """
         # type ignored b/c tuple(MiddlewareOnion) converts to tuple of middlewares
-        all_middlewares: Tuple[Middleware] = tuple(outer_middlewares) + tuple(self.middlewares)  # type: ignore # noqa: E501
+        middlewares: Tuple[Middleware] = tuple(middlewares)  # type: ignore
 
         cache_key = self._request_func_cache[0]
-        if cache_key is None or cache_key != all_middlewares:
+        if cache_key != middlewares:  # type: ignore
             self._request_func_cache = (
-                all_middlewares,
-                self._generate_request_func(w3, all_middlewares),
+                middlewares,
+                combine_middlewares(
+                    middlewares=middlewares,  # type: ignore
```

Comment:
why do we need to ignore the type here? might be worth a comment?

Reply:
This comment actually made me solve the comment above. All these `type: ignores` go away with the `as_tuple_of_middlewares()` method. I could use help naming it though 😆. I like how explicit that name is and I felt like `tupleize` just doesn't explain what it does for a reader trying to move on past the code. Any thoughts there would be appreciated.

Reply:
I can't think of anything better off the top of my head but will let you know if I do!